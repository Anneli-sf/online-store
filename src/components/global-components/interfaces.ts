export interface IComponent {
    render: (id?: number | IProductsData[]) => HTMLDivElement;
}

export interface IRoutes {
    path: string;
    component: IComponent;
}

export interface IProductsData {
    id: number;
    category: string;
    categoryEng: string;
    subcategory: string;
    subcategoryEng: string;
    title: string;
    description: string;
    price: number;
    discount: number;
    stock: number;
    raiting: number;
    images: string[];
}

export interface IStock {
    [key: string]: number;
}

export interface IFilters {
    [key: string]: IProductsData[];
}

export interface IString {
    [key: string]: string[];
}
