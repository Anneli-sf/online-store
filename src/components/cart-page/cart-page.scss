.body {
    background-color: #7f0909;
}
.cart-wrapper {
    position: relative;
    width: 100vw;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: self-start;
    padding-bottom: 100px;
    input {
        background-color: transparent;
    }
    .cart-products {
        background-attachment: fixed;
        width: 60%;
        flex: 0 1 60%;
        margin-right: 15px;
        height: 70vh;
    }

    .block-title {
        font-size: 28px;
    }

    .cart-products {
        .block-header {
            width: 100%;
            display: flex;
            align-items: center;
            color: #ffc500;
            padding: 10px 20px;
            margin-bottom: 10px;
            justify-content: space-between;
            background-color: transparent;
        }
        .quantity,
        .current-page {
            background-color: transparent;
            color: #ffc500;
            pointer-events: none;
            text-align: center;
            font-size: 20px;
        }

        .label,
        .btns-container-switch-pages {
            display: flex;
            align-items: center;
        }

        .title-of-container-switch-btns {
            margin-right: 10px;
        }

        .description-content {
            padding: 10px;
        }

        .btn-switch-page-left,
        .btn-switch-page-right {
            position: relative;
            width: 20px;
            height: 20px;
            background-color: transparent;
            border: none;
            &:hover {
                transform: scale(0.9);
            }
        }

        .btn-switch-page-left::after {
            position: absolute;
            content: url('../../public/icons/arrow-down.svg');
            top: -20px;
            right: -35px;
            transform: rotate(90deg);
        }

        .btn-switch-page-right::after {
            position: absolute;
            content: url('../../public/icons/arrow-down.svg');
            top: -15px;
            right: 20px;
            transform: rotate(-90deg);
            margin-left: 10px;
        }
        .pagination {
            border: 1px solid #ffc500;
            color: #ffc500;
            padding: 5px;
            width: 100px;
            &::placeholder {
                color: #ffc500;
            }
        }
    }

    .stock-value {
        cursor: auto;
        width: 40px;
        text-align: end;
        font-size: 20px;
    }
    .quantity-items-cart {
        text-align: center;
        width: 50px;
        font-size: 20px;
    }

    .cart-products .block-section {
        padding: 10px;
        height: 88%;
        overflow: auto;
        &::-webkit-scrollbar {
            width: 0px;
        }

        .block-header,
        .description-header {
            padding: 5px 0px;
            position: relative;
            background-color: transparent;
            color: #7f0909;
            &::after {
                position: absolute;
                content: '';
                bottom: 0px;
                width: 105%;
                height: 1px;
                background-color: #7f0909;
            }
        }
        .cart-list {
            width: 100%;
            display: flex;
            flex-direction: column;
        }

        .cart-item {
            background-color: #ffc500;
            display: flex;
            justify-content: space-between;
            padding: 5px;
            width: 100%;
            color: #7f0909;
            border-radius: 5px;
            &:not(:last-child) {
                margin-bottom: 10px;
            }
        }

        .cart-item-image {
            width: 130px;
            border-radius: 10px;
        }

        .number-product {
            flex: 1 0 2%;
            margin-right: 10px;
            display: flex;
            justify-content: center;
            align-items: center;
            color: #7f0909;
        }

        .cart-item-image {
            flex: 0 1 20%;
            margin-right: 10px;
            cursor: pointer;
        }

        .description-cart-item {
            margin-right: 10px;
            width: 100%;
            overflow: hidden;
            cursor: pointer;
        }

        .mini-description-item {
            white-space: nowrap;
            text-overflow: ellipsis;
            overflow: hidden;
            max-width: 70%;
            margin: 0 auto;
            margin-bottom: 10px;
        }

        .block-title {
            text-align: center;
            color: #7f0909;
        }

        .values-cart-item {
            position: relative;
            display: flex;
            flex-direction: column;
            justify-content: space-around;
            align-items: center;
            padding: 10px;
            width: 40%;
        }

        .raiting-discount-container {
            display: flex;
            justify-content: space-around;
        }

        .raiting,
        .discount {
            text-decoration: underline;
        }

        .add-del-btns-container {
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        .add-item {
            background-color: transparent;
            border: none;
            position: relative;
            width: 20px;
            height: 20px;
            &:hover {
                transform: scale(0.9);
            }
            &::after {
                position: absolute;
                content: url('../../public/icons/plus.svg');
                width: 20px;
                top: -2px;
                left: -15px;
            }
        }

        .delete-item {
            background-color: transparent;
            border: none;
            position: relative;
            width: 20px;
            height: 20px;
            &:hover {
                transform: scale(0.9);
            }
            &::before {
                position: absolute;
                content: url('../../public/icons/minus.svg');
                width: 20px;
                top: -2px;
                left: 0px;
            }
        }
        .container {
            width: 25%;
        }
        .total-price {
            position: relative;
            font-size: 20px;
            width: 50px;
            text-align: left;
        }
        .span {
            position: absolute;
            right: 0px;
            top: 0px;
            font-size: 20px;
        }
    }

    .summary {
        width: 25%;
        align-self: center;
        border: 1px solid #ffc500;
        border-radius: 5px;

        .block-header {
            width: 100%;
            display: flex;
            align-items: center;
            padding: 10px 20px;
            margin-bottom: 10px;
            color: #ffc500;
        }
        .block-section {
            display: flex;
            flex-direction: column;
            justify-content: space-between;
            align-items: center;
            text-align: center;

            input {
                text-align: center;
                background-color: transparent;
                color: #ffc500;
            }
        }
        .block-section {
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        .quantity-products,
        .total-sum,
        .promecode-input,
        .promo-test {
            color: #ffc500;
            margin-bottom: 10px;
        }

        .promecode-input {
            border: 1px solid #ffc500;
            border-radius: 5px;
        }

        .total-sum-value {
            position: relative;
            font-size: 20px;
        }

        .quantity-products-value {
            font-size: 20px;
        }

        .promecode-input {
            background-color: transparent;
            padding: 5px 20px;
            color: #ffc500;
            &:focus {
                -webkit-box-shadow: 0px 5px 10px 2px rgba(34, 60, 80, 0.2) inset;
                -moz-box-shadow: 0px 5px 10px 2px rgba(34, 60, 80, 0.2) inset;
                box-shadow: 0px 5px 10px 2px rgba(34, 60, 80, 0.2) inset;
            }
            &::placeholder {
                color: #ffc500;
            }
        }

        .form-buy {
            padding: 10px;
        }

        .promo-test {
            font-size: 16px;
            opacity: 0.1;
            &:hover {
                opacity: 1;
            }
        }

        .btn-buy-now {
            background-color: transparent;
            border: 1px solid #ffc500;
            border-radius: 5px;
            color: #ffc500;
            padding: 5px 20px;
            &:hover {
                -webkit-box-shadow: 0px 5px 10px 2px rgba(34, 60, 80, 0.2) inset;
                -moz-box-shadow: 0px 5px 10px 2px rgba(34, 60, 80, 0.2) inset;
                box-shadow: 0px 5px 10px 2px rgba(34, 60, 80, 0.2) inset;
            }
        }
    }
}
