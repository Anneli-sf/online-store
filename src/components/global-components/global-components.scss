@import '../../sass/abstracts/mixins';

.button {
    min-width: 10px;
    min-height: 10px;
    background-color: transparent;
    border: 1px solid white;
    color: white;
    cursor: pointer;
    margin: 5px;
    font-size: 1rem;
}

.input {
    margin: 5px;
    min-width: 150px;
    min-height: 50px;
    background-color: transparent;
    outline: none;
    border: 1px solid white;
    padding: 5px 10px;
    color: white;
    font-size: 1rem;
    cursor: pointer;
}

.label {
    font-size: inherit;
    color: white;
}

.block {
    min-width: 10px;
    min-height: 10px;
    @include flex-column();
    border: 1px solid white;
    color: white;
    font-size: 1em;

    .block-header {
        height: 10%;
        width: 100%;
        background-color: white;
        @include flex-row(center);
        padding: 5px 10px;

        .block-title {
            color: black;
            font-size: 1.17em;
        }
    }

    .block-section {
        padding: 10px 5px;
        height: 100%;
        width: 100%;
        overflow-y: auto;
        overflow-x: hidden;

        .block-text {
            text-align: center;
        }
    }
}
