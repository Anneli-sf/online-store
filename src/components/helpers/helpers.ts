import { IProductsData } from '../data/data';

export const deleteChosenCategory = (currentArr: IProductsData[], chosenCategory: IProductsData[]): IProductsData[] => {
    for (let i = 0; i < chosenCategory.length; i++) {
        const index: number = currentArr.findIndex((item) => item.id === chosenCategory[i].id);
        currentArr.splice(index, 1);
    }
    return currentArr;
};

export const isAlreadyHave = (currentArr: IProductsData[], chosenCategory: IProductsData[]): boolean => {
    return currentArr.some((el) => el.id === chosenCategory[0].id);
};
