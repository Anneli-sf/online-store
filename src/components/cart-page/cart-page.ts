import './cart-page.scss';
import {
    createElement,
    createBlock,
    createButton,
    createLabel,
    createImage,
    createInput,
    createSimpleInput,
    createParagraph,
} from '../global-components/global-components';
import { productsData } from '../data/data';

const listBlock = createElement('ul', 'cart-list') as HTMLUListElement;

export const createCartPage = (): HTMLDivElement => {
    const cartPage = createElement('div', 'cart-wrapper') as HTMLDivElement;
    cartPage.append(createProductsCartBlock(), createSummaryCartBlock());
    return cartPage;
};

function sliceIntoChunks(arr: Array<number>, chunkSize: number) {
    const res = [];
    for (let i = 0; i < arr.length; i += chunkSize) {
        const chunk = arr.slice(i, i + chunkSize);
        res.push(chunk);
    }
    return res;
}

export const createProductsCartBlock = () => {
    const cartBlock = createElement('div', 'cart-products');
    const cartBlockHeader = createElement('div', 'block-header');
    const cartBlockTitle = createElement('p', 'block-title');
    cartBlockTitle.textContent = 'Товары в корзине';

    const cartBlockContent = createElement('div', 'block-section');

    cartBlockHeader.append(cartBlockTitle, quantityProductsOnCart(), switchPagesBlock());

    const fillCartPages = () => {
        const arr = JSON.parse(localStorage.getItem('cartList') as string);
        if (arr.length === 0) {
            return sliceIntoChunks(JSON.parse(localStorage.getItem('cartList') as string), 3);
        } else {
            return sliceIntoChunks(JSON.parse(localStorage.getItem('cartList') as string), 3)[
                +localStorage.getItem('currentPage') - 1
            ];
        }
    };

    const idArray = fillCartPages();
    listBlock.innerHTML = '';
    idArray.forEach((item) => {
        createProductsList(item);
    });

    cartBlockContent.append(listBlock);

    cartBlock.querySelectorAll('.number-product').forEach((item, index) => {
        item.textContent = `${index + 1}`;
    });

    cartBlock.append(cartBlockHeader, cartBlockContent);
    // document.querySelector('.main')?.append(cartBlock);
    return cartBlock;
};

export const createProductsList = (productId: number): HTMLUListElement => {
    listBlock.append(cartProductBlock(productId));

    return listBlock;
};
//----------------------------HEADER
const quantityProductsOnCart = () => {
    const labelQuantityOfProducts = createLabel('Товары: ', 'quantity');
    const inputQuantityOfProducts = createSimpleInput(
        'quantity',
        'number',
        '',
        `${JSON.parse(localStorage.getItem('cartList') as string).length}`
    ) as HTMLInputElement;
    labelQuantityOfProducts.append(inputQuantityOfProducts);

    return labelQuantityOfProducts;
};

if (!localStorage.getItem('currentPage')) {
    localStorage.setItem('currentPage', '1');
}

const switchPagesBlock = () => {
    const pagesSwitchesButtonsBlock = createElement('div', 'btns-container-switch-pages') as HTMLDivElement;

    const currentPage = createSimpleInput(
        'current-page',
        'number',
        '',
        `${localStorage.getItem('currentPage')}`
    ) as HTMLInputElement;

    const textPage = createParagraph('Страница: ', 'title-of-container-switch-btns') as HTMLParagraphElement;

    const buttonSwitchPagesToLeft = createButton('', 'btn-switch-page-left');
    const buttonSwitchPagesToRight = createButton('', 'btn-switch-page-right');

    if (localStorage.getItem('btnLeft') === 'show') {
        buttonSwitchPagesToLeft.style.transform = 'scale(1)';
    } else if (localStorage.getItem('btnLeft') === 'hide') {
        buttonSwitchPagesToLeft.style.transform = 'scale(0)';
    }

    if (localStorage.getItem('btnRight') === 'show') {
        buttonSwitchPagesToRight.style.transform = 'scale(1)';
    } else if (localStorage.getItem('btnRight') === 'hide') {
        buttonSwitchPagesToRight.style.transform = 'scale(0)';
    }

    pagesSwitchesButtonsBlock.append(textPage, buttonSwitchPagesToLeft, currentPage, buttonSwitchPagesToRight);

    return pagesSwitchesButtonsBlock;
};

//----------------------------/HEADER

const cartProductBlock = (productId?: number) => {
    const productCartBlock = createElement('li', 'cart-item') as HTMLLIElement;
    productCartBlock.id = `${productId}`;
    productCartBlock.append(
        numberOfProductBlock(productId),
        imageProductBlock(productId),
        productsDescriptionBlock(productId),
        productsValuesBlock(productId)
    );

    return productCartBlock;
};

const numberOfProductBlock = (productId: number) => {
    const numberOfProduct = createElement('p', 'number-product') as HTMLParagraphElement;
    numberOfProduct.textContent = `${JSON.parse(localStorage.getItem('cartList') as string).indexOf(productId) + 1}`;
    return numberOfProduct;
};

const imageProductBlock = (productId: number): HTMLImageElement => {
    const image = createImage(
        productsData[productId].images[0],
        productsData[productId].title,
        'cart-item-image'
    ) as HTMLImageElement;
    return image;
};

const productsDescriptionBlock = (productId: number) => {
    const descriptionBlock = createElement('div', 'description-cart-item') as HTMLDivElement;
    const descriptionHeader = createElement('div', 'description-header') as HTMLDivElement;
    const descriptionTitle = createParagraph(productsData[productId].title, 'block-title') as HTMLParagraphElement;

    const descriptionContent = createElement('div', 'description-content') as HTMLDivElement;

    const miniDescription = createParagraph(
        productsData[productId].description,
        'mini-description-item'
    ) as HTMLParagraphElement;

    const raitingDiscountContainer = createElement('div', 'raiting-discount-container') as HTMLDivElement;

    const raiting = createParagraph('Рейтинг: ', 'raiting') as HTMLParagraphElement;
    const discount = createParagraph('Скидка: ', 'discount') as HTMLParagraphElement;

    const raitingValue = createSimpleInput('raiting-value', '', '', `${productsData[7].raiting}`) as HTMLInputElement;
    raitingValue.setAttribute('readonly', 'true');

    const discountValue = createSimpleInput(
        'discount-value',
        '',
        '',
        `${productsData[7].discount}%`
    ) as HTMLInputElement;
    discountValue.setAttribute('readonly', 'true');

    raiting.append(raitingValue);
    discount.append(discountValue);
    raitingDiscountContainer.append(raiting, discount);

    descriptionHeader.append(descriptionTitle);
    descriptionContent.append(miniDescription, raitingDiscountContainer);
    descriptionBlock.append(descriptionHeader, descriptionContent);

    return descriptionBlock;
};

const productsValuesBlock = (productId: number) => {
    const valuesBlock = createElement('div', 'values-cart-item') as HTMLDivElement;
    const stock = createParagraph('Доступно: ', 'stock') as HTMLParagraphElement;

    if (!localStorage.getItem(`stock_${productId}`)) {
        localStorage.setItem(`stock_${productId}`, productsData[productId].stock - 1);
    }

    const stockValue = createSimpleInput(
        'stock-value',
        'number',
        '',
        `${localStorage.getItem(`stock_${productId}`)}`,
        '',
        '',
        '',
        `${productId}`
    ) as HTMLInputElement;
    stockValue.setAttribute('readonly', 'true');
    const shtuk = createElement('span', 'stock-text') as HTMLDivElement;
    shtuk.textContent = `шт.`;

    if (!localStorage.getItem(`quantityProduct_${productId}`)) {
        localStorage.setItem(`quantityProduct_${productId}`, '1');
    }

    const quantityOfItemsInCart = createSimpleInput(
        'quantity-items-cart',
        'number',
        '',
        +localStorage.getItem(`quantityProduct_${productId}`)
    ) as HTMLInputElement;
    quantityOfItemsInCart.setAttribute('readonly', 'true');

    const addAndDelItemsButtonsContainer = createElement('div', 'add-del-btns-container') as HTMLDivElement;
    const addItemButton = createButton('', 'add-item') as HTMLButtonElement;
    addItemButton.id = `${productId}`;
    const deleteItemButton = createButton('', 'delete-item') as HTMLButtonElement;
    deleteItemButton.id = `${productId}`;

    const container = createElement('div', 'container') as HTMLDivElement;
    const totalPrice = createElement('input', 'total-price') as HTMLInputElement;
    totalPrice.readOnly = true;
    totalPrice.id = `${productId}`;
    if (!localStorage.getItem(`price_${productId}`)) {
        localStorage.setItem(`price_${productId}`, `${productsData[productId].price}`);
        totalPrice.value = `${productsData[productId].price}`;
    } else {
        (totalPrice.value as string | null) = localStorage.getItem(`price_${productId}`);
    }
    const span = createElement('span', 'span');
    span.textContent = '$';
    container.append(totalPrice, span);

    stock.append(stockValue, shtuk);
    addAndDelItemsButtonsContainer.append(deleteItemButton, quantityOfItemsInCart, addItemButton);
    valuesBlock.append(stock, addAndDelItemsButtonsContainer, container);

    return valuesBlock;
};

//-----------------------------------------------------

//------------------STOCK and QUANTITY------------------
export const createSummaryCartBlock = () => {
    const summaryBlock = createBlock('summary', 'Итого') as HTMLElement;

    const summarySectionBlock = summaryBlock.querySelector('.block-section') as HTMLElement;

    const quantityOfPoducts = createParagraph('Товары, шт: ', 'quantity-products') as HTMLParagraphElement;

    let totalStock = JSON.parse(localStorage.getItem('cartList') as string).length;

    if (localStorage.getItem('totalStock')) {
        totalStock = +localStorage.getItem('totalStock');
    }

    const quantityOfPoductsValue = createSimpleInput(
        'quantity-products-value',
        '',
        '',
        `${totalStock}`,
        '',
        '',
        true
    ) as HTMLInputElement;
    const totalSum = createParagraph('Сумма: ', 'total-sum') as HTMLParagraphElement;

    let totalPrice = JSON.parse(localStorage.getItem('cartItems') as string)
        .flat()
        .reduce((acc, curr) => acc + curr.price, 0);

    if (localStorage.getItem('totalPrice')) {
        totalPrice = +localStorage.getItem('totalPrice');
    }
    const totalSumValue = createSimpleInput(
        'total-sum-value',
        '',
        '',
        `${totalPrice}`,
        '',
        '',
        true
    ) as HTMLInputElement;

    summarySectionBlock.append(
        quantityOfPoducts,
        quantityOfPoductsValue,
        totalSum,
        totalSumValue,
        formOfBuyProductsBlock()
    );
    // document.querySelector('.main')?.append(summaryBlock);

    return summaryBlock;
};

const formOfBuyProductsBlock = () => {
    const buyForm = createInput('promecode-input', 'text', 'Введите промокод') as HTMLFormElement; //'form-buy'

    const promoTest = createParagraph("Попробуйте: 'Гарри', 'Поттер'", 'promo-test') as HTMLParagraphElement;
    const buttonBuyNow = createButton('Купить сейчас', 'btn-buy-now');

    buyForm.append(promoTest, buttonBuyNow);

    return buyForm;
};
