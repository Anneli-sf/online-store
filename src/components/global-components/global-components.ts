import './global-components.scss';

// T extends keyof HTMLElementTagNameMap

export const createElement = (elTag: string, elClassName: string): HTMLElement => {
    const element = document.createElement(`${elTag}`);
    element.classList.add(elClassName);
    return element;
};

export const createButton = (buttonText: string, classButton?: string): HTMLButtonElement => {
    const button = createElement('button', 'button') as HTMLButtonElement;
    classButton ? button.classList.add(classButton) : undefined;
    button.innerText = buttonText;
    return button;
};

export const createInput = (
    inputClass: string,
    inputType: string,
    inputPlaceHolder?: string,
    inputId?: string,
    inputName?: string,
    inputValue?: string,
    minValue?: string,
    maxValue?: string
): HTMLFormElement => {
    const form = createElement('form', 'form') as HTMLFormElement;

    const input = createElement('input', `${inputClass}`) as HTMLInputElement;
    input.type = inputType;
    inputId ? (input.id = inputId) : undefined;
    inputName ? (input.name = inputName) : undefined;
    inputPlaceHolder ? (input.placeholder = inputPlaceHolder) : undefined;
    inputValue ? (input.value = inputValue) : undefined;
    minValue ? (input.min = minValue) : undefined;
    maxValue ? (input.max = maxValue) : undefined;

    form.append(input);
    return form;
};

export const createSimpleInput = (
    inputClass: string,
    inputType: string,
    inputPlaceHolder?: string,
    inputValue?: string,
    minValue?: string,
    maxValue?: string,
    readOnly?: boolean,
    id?: string
): HTMLInputElement => {
    const input = createElement('input', `${inputClass}`) as HTMLInputElement;
    input.type = inputType;
    inputPlaceHolder ? (input.placeholder = inputPlaceHolder) : undefined;
    inputValue ? (input.value = inputValue) : undefined;
    minValue ? (input.min = minValue) : undefined;
    maxValue ? (input.max = maxValue) : undefined;
    if (readOnly) input.readOnly = readOnly;
    if (id) input.id = id;

    return input;
};

export const createLabel = (labelText: string, labelClass: string, labelName?: string): HTMLLabelElement => {
    const label = createElement('label', `${labelClass}`) as HTMLLabelElement;
    labelName ? label.setAttribute('for', labelName) : undefined;
    label.innerText = labelText;
    return label;
};

export const createBlock = (
    blockClass: string,
    titleText?: string,
    headerClass?: string,
    sectionClass?: string
): HTMLElement => {
    const container = createElement('div', 'block') as HTMLElement;
    container.classList.add(blockClass);

    const blockHeader = createElement('div', 'block-header') as HTMLElement;
    headerClass ? blockHeader.classList.add(headerClass) : undefined;
    const title = createElement('p', 'block-title');
    if (titleText) title.innerHTML = `${titleText}`;

    const blockSection: HTMLElement = createElement('div', 'block-section');
    sectionClass ? blockSection.classList.add(sectionClass) : undefined;

    blockHeader.append(title);
    container.prepend(blockHeader);
    container.append(blockSection);

    return container;
};

export const createBlockWithText = (
    blockClass: string,
    pClass: string,
    text: string | number,
    headerClass?: string,
    titleText?: string,
    sectionClass?: string
    // productId?: number
): HTMLElement => {
    const container: HTMLElement = createElement('div', 'block');
    container.classList.add(blockClass);

    const blockHeader: HTMLElement = createElement('div', 'block-header');
    headerClass ? blockHeader.classList.add(headerClass) : undefined;
    const title: HTMLElement = createElement('p', 'block-title');
    if (titleText) title.innerHTML = `${titleText}`;

    const blockSection: HTMLElement = createElement('div', 'block-section');
    sectionClass ? blockSection.classList.add(sectionClass) : undefined;

    const blockText: HTMLElement = createElement('p', 'block-text');
    pClass ? blockText.classList.add(pClass) : undefined;
    text ? (blockText.textContent = '' + text) : undefined; //передается число или строка

    blockSection.append(blockText);
    blockHeader.append(title);
    container.prepend(blockHeader);
    container.append(blockSection);

    return container;
};

export const createImage = (imgSrc: string, imgAlt: string, imgClass?: string, id?: string): HTMLImageElement => {
    const image = createElement('img', 'image') as HTMLImageElement;
    imgClass ? image.classList.add(imgClass) : undefined;
    image.src = imgSrc;
    image.alt = imgAlt;
    image.id = id as string;
    return image;
};

export const createLink = (linkHref: string, linkClass: string, linkText?: string): HTMLLinkElement => {
    const link = createElement('a', linkClass) as HTMLLinkElement;
    link.href = linkHref;
    linkText ? (link.textContent = linkText) : undefined;

    return link;
};

export const createParagraph = (parText: string | number, parClass: string): HTMLParagraphElement => {
    const text = createElement('p', parClass) as HTMLParagraphElement;
    text.innerHTML = `${parText}`;
    return text;
};
