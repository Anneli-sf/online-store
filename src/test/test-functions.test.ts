import { createEmptyPage, sliceIntoChunks } from '../components/cart-page/cart-page';
import { productsData } from '../components/data/data';
import { stateFilters } from '../components/main-section/main-section-index';
import { isAlreadyHave, getMinPrice, getMaxPrice, getMinAmount, getMaxAmount } from '../components/helpers/helpers';
import { IProductsData } from '../components/global-components/interfaces';
import { createElement } from '../components/global-components/global-components';

const testArray: IProductsData[] = [
    {
        id: 0,
        category: 'Магические существа',
        categoryEng: 'magical-animals',
        subcategory: 'Отрицательные существа',
        subcategoryEng: 'negative-animals',
        title: 'Василиск',
        description:
            'Огромный змей, известный также как «король змей», живущий не одну сотню лет. Очень сильное магическое животное.',
        price: 305,
        discount: 10,
        stock: 5,
        raiting: 4.95,
        images: ['assets/images-storage/vasilisk/vasil1.jpg', 'assets/images-storage/vasilisk/vasil2.jpg'],
    },
    {
        id: 1,
        category: 'Магические предметы',
        categoryEng: 'magic-items',
        subcategory: 'Оружие',
        subcategoryEng: 'weapon',
        title: 'Меч Гриффиндора',
        description:
            'Магический предмет исключительной силы. Это одноручный меч (34 дюйма), созданный из чистого серебра, и достаточно лёгкий, чтобы его мог поднять даже подросток. Рукоять меча отделана рубинами, на лезвии клинка видна гравировка «Годрик Гриффиндор».',
        price: 10,
        discount: 10,
        stock: 3,
        raiting: 4.05,
        images: [
            'assets/images-storage/mech_Griffindora/grif1.png',
            'assets/images-storage/mech_Griffindora/grif2.jpg',
        ],
    },
    {
        id: 2,
        category: 'Одежда и аксессуары',
        categoryEng: 'clothes-accessories',
        subcategory: 'Одежда',
        subcategoryEng: 'clothes',
        title: 'Мантия-невидимка',
        description:
            'Волшебная мантия из очень лёгкой «серебристой» ткани, делающая невидимым того, кто её надевает. Обычные мантии-невидимки служат очень недолго и их производство очень дорогостоящее. У Гарри была особая мантия-невидимка, так как являлась одним из трех Даров Смерти.',
        price: 15,
        discount: 9,
        stock: 8,
        raiting: 3.98,
        images: [
            'assets/images-storage/mantiya_nevidimka/mant1.jpg',
            'assets/images-storage/mantiya_nevidimka/mant2.png',
        ],
    },
];

describe('createEmptyPage', () => {
    test('Block is created!', () => {
        expect(createEmptyPage()).toMatchSnapshot();
    });
});

describe('sliceIntoChunks', () => {
    test('Array divided into parts correctly!', () => {
        expect(sliceIntoChunks([1, 2, 3, 4], 2)).toEqual(
            expect.arrayContaining([
                [1, 2],
                [3, 4],
            ])
        );
    });
    test('Array divided into parts correctly!', () => {
        expect(sliceIntoChunks([1, 2, 3, 4, 5], 3)).toEqual(
            expect.arrayContaining([
                [1, 2, 3],
                [4, 5],
            ])
        );
    });
});

describe('stateFilters', () => {
    const element = createElement('input', 'sort__input') as HTMLInputElement;
    element.value = 'element';
    test('Return slash!', () => {
        expect(stateFilters([], [], '', '', '', '', productsData, element)).toEqual('');
    });
    test('Return only category!', () => {
        expect(stateFilters(['abc'], [], '', '', '', '', testArray, element)).toEqual(
            '?category=abc&&price=↕&stock=↕&search=element'
        );
    });
    test('Return search value!', () => {
        expect(stateFilters([], [], '', '', '', '', testArray, element)).toEqual('?&&price=↕&stock=↕&search=element');
    });
    test('Return category + subcategory + stock + price!', () => {
        expect(stateFilters(['abc'], ['cba'], '0', '100', '0', '100', testArray, element)).toEqual(
            '?category=abc&subcategory=cba&price=0↕100&stock=0↕100&search=element'
        );
    });
});

describe('isAlreadyHave', () => {
    test('It is true!', () => {
        expect(isAlreadyHave(productsData, testArray)).toBeTruthy();
    });
});

describe('getMinPrice', () => {
    test('It is a min price!', () => {
        expect(getMinPrice(testArray)).toEqual(10);
    });
});

describe('getMaxPrice', () => {
    test('It is a max price!', () => {
        expect(getMaxPrice(testArray)).toEqual(305);
    });
});

describe('getMinAmount', () => {
    test('It is a min stock!', () => {
        expect(getMinAmount(testArray)).toEqual(3);
    });
});

describe('getMaxAmount', () => {
    test('It is a max stock!', () => {
        expect(getMaxAmount(testArray)).toEqual(8);
    });
});
