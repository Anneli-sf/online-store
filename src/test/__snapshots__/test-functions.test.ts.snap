// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEmptyPage Block is created! 1`] = `
<div
  class="empty"
>
  <p
    class="empty-text"
  >
    Корзина пуста...
  </p>
</div>
`;
